src/v/pandaproxy/schema_registry/util.h

```diff
@@ -47,4 +55,57 @@ result<schema_definition> make_schema_definition(std::string_view sv) {
       ss::sstring{str_buf.GetString(), str_buf.GetSize()}};
 }
 
+class client_fetcher final : public model::record_batch_reader::impl {
+    using storage_t = model::record_batch_reader::storage_t;
+
+public:
+    client_fetcher(
+      kafka::client::client& client,
+      model::topic_partition tp,
+      model::offset first,
+      model::offset last)
+      : _client{client}
+      , _tp{std::move(tp)}
+      , _next_offset{first}
+      , _last_offset{last}
+      , _batch_reader{} {}
+
+    // Implements model::record_batch_reader::impl
+    bool is_end_of_stream() const final { return _next_offset >= _last_offset; }
+
+    // Implements model::record_batch_reader::impl
+    ss::future<storage_t>
+    do_load_slice(model::timeout_clock::time_point t) final {
+        if (!_batch_reader || _batch_reader->is_end_of_stream()) {
+            vlog(plog.debug, "Schema registry: fetch offset: {}", _next_offset);
+            auto res = co_await _client.fetch_partition(
+              _tp, _next_offset, 1_MiB, t - model::timeout_clock::now());
+            vlog(plog.debug, "Schema registry: fetch result: {}", res);
+            vassert(res.begin() != res.end(), "Can this fail?");
+            _batch_reader = std::move(res.begin()->partition_response->records);
+            vassert(++res.begin() == res.end(), "Can this fail?");
+        }
+        auto ret = co_await _batch_reader->do_load_slice(t);
+        using data_t = model::record_batch_reader::data_t;
+        vassert(std::holds_alternative<data_t>(ret), "Can this fail?");
+        auto& data = std::get<data_t>(ret);
+        vassert(!data.empty(), "Can this fail?");
```

Comment:
since it looks like this is coming from a fetch request, it probably can be empty in some cases.

Reply:
Fixed